nix/supervisord-cluster/profiles.jq

```diff
@@ -0,0 +1,88 @@
+## classified into sections:
+##
+##  - genesis
+##  - generator
+##  - node
+##  - tolerances (ranges of acceptable properties, given above parameters)
+##
+## When combined with cluster composition (an extract from topology, see the
+## definition below) and service parameters (non-crucial things like state directory,
+## port ranges, etc), profiles ought to completely specify cluster parameters,
+## allowing genesis and all necessary configuration files to be generated.
+##
+## Profiles themselves are layered as follows:
+##
+##  - era-dependent defaults for the aforementioned sections:
+##    - profiles/defaults.jq
+##
+##  - overlayed with generated profile variants + ad-hoc profiles:
+##    - profiles/variants.jq and profiles/adhoc.jq
+##
+##  - each then further overlayed with derived parameters, computed from the above:
+##    - profiles/derived.jq
+##
+## Profiles variants are generated as a cartesian product of variations
+## of the three axes: genesis, generator and node.
+## These generated profiles are assigned computed names, as per the
+## profile_name() function in 'profiles/derived.jq'.
+##
+## Cluster composition must have the following structure:
+##  { n_hosts:          INT
+##  , n_bft_hosts:      INT
+##  , n_singular_hosts: INT
+##  , n_dense_hosts:    INT
+##  }
+##  ..where n_hosts must be equal to a sum of the rest.
+##
+## Names of non-adhoc profiles are computed by the 'profile_name' function in
+## profiles/derived.jq, with era suffix appended, and have the following structure:
+##
+##   k${n_pools}
+##    -[${dense_pool_density}ppn]
+##    -${epochs}ep
+##    -${utxo}kU
+##    -${delegators}kD
+##    -[${tps}tps]
+##    -[${max_block_size}blk]
+##    -[${add_tx_size}b]
+##    -[${inputs_per_tx}i]
+##    -[${outputs_per_tx}o]
+##
+## ..where [] denote optionals, that are only included if the profile
+##   deviates from profile defaults.
+##
+## Testable by:
+##
+##   nix-build -A profiles       ## or simply:  make profiles
+##
+## ..which simply calls ./profiles.nix with {} params.
+##
+
+include "topology"    { search: "profiles" };
+include "defaults"    { search: "profiles" };
+include "adhoc"       { search: "profiles" };
+include "variants"    { search: "profiles" };
+include "derived"     { search: "profiles" };
+
+def profiles($era; $mcompo; $topo; $extra_profiles):
+    ($mcompo // topology_composition($topo // {}) // {}) as $compo
+
+  ## Profiles are variants + custom (or aux) profiles:
+  | all_profile_variants + adhoc_profiles + $extra_profiles
+
+  | map (## Each profile extends defaults:
+         era_defaults($era) * .
+
+         ## Profiles can define their own cluster composition.
+         | . * { composition: (.composition // $compo) }
+
+         ## Compute the derived params.
+         | add_derived_params
+
+         ## Finally, assembly into a dictionary..
+         | { "\(.name)":
+               ## ..and cleanup:
+               . | delpaths ([["generator", "epochs"]])}
+        )
+  | add;
```

Comment:
I've never seen `jq` files before.  This is cool.  😁

Reply:
As soon as this stops evolving too fast, we'll probably migrate them to Haskell : -)